### PR TITLE
[MS-334] Dataset Converter

### PR DESCRIPTION
**Pull request by imda-normanchia**
## Description

This feature allows user to convert their existing dataset in csv format to the json in our format. The requirement from them would be to have a csv file that contains 2 column, target and input. The csv converter looks for this 2 column in the csv file and convert it to the dataset shape of our own. 

This feature also allows user to download datasets from huggingface using moonshot and it will convert the downloaded dataset from huggingface into the dataset shape of our own. 

## Motivation and Context

This allows user to easily convert their dataset or download new dataset to use with moonshot.
 
## How to Test

[Provide clear instructions on how to test and verify the changes introduced by this pull request, including any specific unit tests you have created to demonstrate your changes.]

## Checklist

Please check all the boxes that apply to this pull request using "x":

- [x]  I have tested the changes locally and verified that they work as expected.
- [ ]  I have added or updated the necessary documentation (README, API docs, etc.).
- [x]  I have added appropriate unit tests or functional tests for the changes made.
- [x]  I have followed the project's coding conventions and style guidelines.
- [ ]  I have rebased my branch onto the latest commit of the main branch.
- [ ]  I have squashed or reorganized my commits into logical units.
- [x]  I have added any necessary dependencies or packages to the project's build configuration.
- [x]  I have performed a self-review of my own code.
- [x]  I have read, understood and agree to the Developer Certificate of Origin below, which this project utilises.

## Screenshots (if applicable)

[If the changes involve visual modifications, include screenshots or GIFs that demonstrate the changes.]

## Additional Notes

[Add any additional information or context that might be relevant to reviewers.]

<details>
  <summary>Developer Certificate of Origin</summary>

 ```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
 ```
</details>

### CODE REVIEW

File: moonshot/src/api/api_dataset.py [hunk 1]
```diff
@@ -44,3 +45,36 @@ def api_get_all_datasets_name() -> list[str]:
     """
     datasets_name, _ = Dataset.get_available_items()
     return datasets_name
+
+
+def api_create_datasets(
+    name: str, description: str, reference: str, license: str, method: str, **kwargs
```
Comment (imda-lionelteo) [inline moonshot/src/api/api_dataset.py:51]: what is this method:str

File: moonshot/src/api/api_dataset.py [hunk 2]
```diff
@@ -44,3 +45,36 @@ def api_get_all_datasets_name() -> list[str]:
...
+    DatasetArguments object with these details and an empty id. The id is left empty because it will be generated
+    from the name during the creation process. The function then calls the Dataset's create method to
+    create the new dataset.
+
+    Args:
+        name (str): The name of the new dataset.
+        description (str): A brief description of the new dataset.
+        reference (str): A reference link for the new dataset.
+        license (str): The license of the new dataset.
+        kwargs: Additional keyword arguments for the Dataset's create method.
```
Comment (imda-lionelteo) [inline moonshot/src/api/api_dataset.py:66]: method doc is not here

File: moonshot/integrations/cli/common/dataset.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+from rich.console import Console
+
+from moonshot.api import (
+    api_create_datasets,
+)
+
+console = Console()
+
+def add_dataset(args) -> None:
+    try:
```
Comment (imda-lionelteo) [inline moonshot/integrations/cli/common/dataset.py:13]: docstrings

File: moonshot/src/datasets/dataset.py [hunk 4]
```diff
@@ -17,6 +20,118 @@ class Dataset:
...
+                "name": ds_args.name,
+                "description": ds_args.description,
+                "reference": ds_args.reference,
+                "license": ds_args.license,
+                "examples": examples,
+            }
+
+            # Write as JSON output
+            file_path = Storage.create_object(EnvVariables.DATASETS.name, ds_id, ds_info, "json")
+            print(file_path)
```
Comment (imda-lionelteo) [inline moonshot/src/datasets/dataset.py:85]: is this expected?

File: moonshot/src/datasets/dataset.py [hunk 5]
```diff
@@ -17,6 +20,118 @@ class Dataset:
...
+            return file_path
+
+        except Exception as e:
+            print(f"Failed to create dataset: {str(e)}")
```
Comment (imda-lionelteo) [inline moonshot/src/datasets/dataset.py:89]: can we use logger instead

File: tests/unit-tests/web_api/test_routes/test_routes_bookmark.py [hunk 6]
```diff
@@ -68,7 +72,7 @@ def test_get_all_bookmarks(test_client, mock_bookmark_service, bookmark_id, mock
 
     # Make the GET request to the API
     response = test_client.get(f"/api/v1/bookmarks?id={bookmark_id}" if bookmark_id else "/api/v1/bookmarks")
-
+    print(response.json())
```
Comment (imda-lionelteo) [inline tests/unit-tests/web_api/test_routes/test_routes_bookmark.py:75]: is this expected?

File: moonshot/src/api/api_dataset.py (hunk 1 above)
Comment (imda-normanchia) [reply moonshot/src/api/api_dataset.py:51]: whether to convert dataset from hugging face or csv

File: moonshot/src/datasets/dataset.py [hunk 7]
```diff
@@ -17,6 +20,118 @@ class Dataset:
...
+        where each dictionary represents a row in the CSV file.
+
+        Args:
+            csv_file (str): The file path to the CSV file.
+
+        Returns:
+            list[dict]: A list of dictionaries representing the CSV data.
+        """
+        df = pd.read_csv(csv_file)
+        data = df.to_dict("records")
```
[the remark this answers is not in the document]
Comment (imda-normanchia) [reply moonshot/src/datasets/dataset.py:107]: the records is a key for pandas framework to convert the csv to rows of dictionary

**Review by imda-lionelteo — APPROVED** [review]

**Comment by imda-benedictlee** [comment]
Link to Jira Ticket: https://imda-dsl.atlassian.net/browse/MS-334?atlOrigin=eyJpIjoiYWM2YjAxODllNzUwNDY3ZThiOTU4YTFlNjBkZDA1NDAiLCJwIjoiaiJ9

**Comment by imda-benedictlee** [comment]
Tested. Working as expected.

Create Dataset via CSV:
Screenshot 2024-07-31 at 12 50 53 PM

Screenshot 2024-07-31 at 12 53 07 PM

Screenshot 2024-07-31 at 12 53 10 PM

Create Dataset via CSV using the same name:
Screenshot 2024-07-31 at 1 01 22 PM

Create Dataset via Hugging Face:
Screenshot 2024-07-31 at 12 58 31 PM

Screenshot 2024-07-31 at 12 59 02 PM

Screenshot 2024-07-31 at 12 59 06 PM

Create Dataset via Hugging Face using the same name:
Screenshot 2024-07-31 at 1 02 32 PM

**Pull request merged**